src/components/ThemeProvider/ThemeProvider.tsx

```diff
@@ -19,28 +21,73 @@ export function ThemeProvider({
   children,
 }: ThemeProviderProps) {
   const {unstableGlobalTheming = false} = useFeatures();
+
+  const parentContext = useContext(ThemeContext);
+  const isParentThemeProvider = parentContext === undefined;
+  const parentMode = parentContext && parentContext.mode && parentContext.mode;
+  const parentColors =
+    parentContext &&
+    parentContext.UNSTABLE_colors &&
+    parentContext.UNSTABLE_colors;
+
+  const {UNSTABLE_colors, mode, ...rest} = themeConfig;
+
+  const childShouldInheritParentColors =
+    !isParentThemeProvider && mode !== undefined && mode !== parentMode;
+
+  const defaultColors = {
+    surface: UNSTABLE_Color.Surface,
+    onSurface: UNSTABLE_Color.OnSurface,
+    interactive: UNSTABLE_Color.Interactive,
+    neutral: UNSTABLE_Color.Neutral,
+    primary: UNSTABLE_Color.Primary,
+    critical: UNSTABLE_Color.Critical,
+    warning: UNSTABLE_Color.Warning,
+    highlight: UNSTABLE_Color.Highlight,
+    success: UNSTABLE_Color.Success,
+    decorative: UNSTABLE_Color.Decorative,
+  };
+
+  const processedThemeConfig: ThemeConfig = {
+    ...rest,
+    ...{mode: mode || parentMode},
+    UNSTABLE_colors: {
+      ...(isParentThemeProvider && defaultColors),
+      ...(childShouldInheritParentColors && parentColors),
+      ...UNSTABLE_colors,
+    },
+  };
+
   const customProperties = useMemo(
-    () => buildCustomProperties(themeConfig, unstableGlobalTheming),
-    [unstableGlobalTheming, themeConfig],
+    () =>
+      buildCustomProperties(
+        processedThemeConfig,
+        unstableGlobalTheming,
+        isParentThemeProvider ? Tokens : undefined,
+      ),
+    [isParentThemeProvider, processedThemeConfig, unstableGlobalTheming],
   );
+
   const theme = useMemo(
     () =>
       buildThemeContext(
-        themeConfig,
+        processedThemeConfig,
         unstableGlobalTheming ? customProperties : undefined,
       ),
-    [customProperties, themeConfig, unstableGlobalTheming],
+    [customProperties, processedThemeConfig, unstableGlobalTheming],
   );
 
-  // We want these values to be `null` instead of `undefined` when not set.
+  // We want these values to be empty string instead of `undefined` when not set.
   // Otherwise, setting a style property to `undefined` does not remove it from the DOM.
-  const backgroundColor = customProperties['--p-surface-background'] || null;
-  const color = customProperties['--p-text-on-surface'] || null;
+  const backgroundColor = customProperties['--p-surface-background'] || '';
+  const color = customProperties['--p-text-on-surface'] || '';
```

Comment:
Just realized in the case where the `mode` is different than the parent, the `color` needs to be set on the wrapping `div`, too. Otherwise, all of the text that inherits the text color will be the wrong color.

<img src="https://user-images.githubusercontent.com/344839/70598772-70d5d280-1ba0-11ea-9ed4-1ac2ee20b638.png" width="200" />

Reply:
Makes sense, but I think we should do this with CSS on the ThemeProvider node. That avoids the additional complexity a JavaScript solution would add.